CRADLE/CorrectBiasStored/correctReadCounts.py

```diff
@@ -0,0 +1,151 @@
+	# covariates have been precomputed for
+	if shearStart < 1:
+		fragRegionStart = SONICATION_SHEAR_BIAS_OFFSET + START_INDEX_ADJUSTMENT
+		analysisStart = max(analysisStart, fragRegionStart)
+
+	if shearEnd > chromoEnd:
+		fragRegionEnd = chromoEnd - SONICATION_SHEAR_BIAS_OFFSET
+		analysisEnd = min(analysisEnd, fragRegionEnd)
+
+	return (analysisStart, analysisEnd)
+
+def writeCorrectedReads(outFile, sectionCount, starts, ends, values):
+	for i in range(sectionCount):
+		outFile.write(struct.pack(CORRECTED_RC_TEMP_FILE_STRUCT_FORMAT, starts[i], ends[i], values[i]))
+
+def correctReadCount(regions, chromoEnds, covariates, trainingBWName, bwNames, scalers, COEFs, COEF_HIGHRCs, highRC, minFragFilterValue, binsize, outputDir):
+	meanMinFragFilterValue = int(np.round(minFragFilterValue / len(bwNames)))
+
+	for chromoRegionData in regions:
+		chromo, chromoId, chromoRegions = chromoRegionData
+
+		# Align regions to the covariate file boundaries, generate the "overall" indices used later and load up the training read counts
+		# all in one loop
+		overallIndices = []
+		adjustedChromoRegions = []
+		with pyBigWig.open(trainingBWName) as bwFile:
+			trainingReadCounts = np.zeros(bwFile.chroms(chromo), dtype=np.float32)
+			for region in chromoRegions:
+				# Align the region to covariate file boundaries
+				start, end = alignCoordinatesToCovariateFileBoundaries(region, chromoEnds, covariates.fragLen)
+				adjustedChromoRegions.append((start, end))
+
+				# generate the "overall" index of locations with total read counts > minFragFilterValue
+				overallIndices.append(selectOverallIdx(chromo, start, end, bwNames, minFragFilterValue))
+
+				# load the training read counts
+				if pyBigWig.numpy == 1:
+					trainingReadCounts[start:end] = bwFile.values(chromo, start, end, numpy=True)
+				else:
+					trainingReadCounts[start:end] = bwFile.values(chromo, start, end)
+
+			trainingReadCounts[np.isnan(trainingReadCounts)] = 0.0
+
+		del chromoRegions # We don't need this anymore and should break if we use it
+
+		for bwName, scaler, COEF, COEF_HIGHRC in zip(bwNames, scalers, COEFs, COEF_HIGHRCs):
+			bwFile = pyBigWig.open(bwName)
+			correctedReadCountOutputFile = open(outputCorrectedTmpFile(outputDir, chromo, chromoId, bwName), "wb")
+			covariateFile = h5py.File(covariates.covariateFileName(chromo), "r")
+			covariateValues = covariateFile['covari']
+
+			for region, overallIdx in zip(adjustedChromoRegions, overallIndices):
+				analysisStart, analysisEnd = region
+
+				###### GET POSITIONS WHERE THE NUMBER OF FRAGMENTS > MIN_FRAGNUM_FILTER_VALUE
+				if pyBigWig.numpy == 1:
+					rcArr = bwFile.values(chromo, analysisStart, analysisEnd, numpy=True).astype(np.float64)
+				else:
+					rcArr = np.array(bwFile.values(chromo, analysisStart, analysisEnd))
+				rcArr[np.isnan(rcArr)] = 0.0
+				readCountIdx = selectReplicateIdx(rcArr, overallIdx, meanMinFragFilterValue)
```

Comment:
I am still debating with myself which one is the right approach. The previous version filtered the same regions across all samples—if one sample does not have enough coverage for a certain genomic region, that region was filtered out across all samples. 

The advantage of filtering the same regions across all samples is to avoid false positives in detecting peaks.
For example,
```
In region R, 
 the read count of rep1 with treatment A = 20
  the read count of rep2 with treatment A = 1
```
If `meanMinFragFilterValue=5`, the region R gets filtered out only in rep2 in the current version while the region R gets filtered out in both replicates. 

If the region R gets filtered out only in rep2, the callPeak subcommand (or other peak callers) might consider the resulting corrected read counts as zero and perform a statistical testing in that region R—which might result in a false positive. 

However, we can lose a lot of statistical power from filtering the same regions across the samples. For example, we have multiple treatment—just like Graham's data. If the samples from one of the treatment have particularly bad library quality and low coverage, we lose the opportunity to test many regions for other treatments. 

 I think the best way to do this is filtering the same regions across the replicates within the same biological condition. However, we currently have no argument that we can guess the biological condition of each sample. What I would suggest instead is 
- filtering the same regions in the single run
- putting in the README that we strongly suggest running the algorithm separately for each biological condition but using the same `-rngSeed`.

Alternatively, we can add another argument to know the biological conditions of the samples. 

What do you think?

Reply:
I think I do cover this situation -- That's why overallIndices is calculated on line 62: https://github.com/ReddyLab/CRADLE/pull/78/files/c23ab4c9293eef8eb62c1aa1013162a071bbc9e9#diff-1fc2cff4e79bf185f9b94efbde101e4cd33c6b1d310d47459d05367f6fe0355cR62

and the `selectOverallIdx` function https://github.com/ReddyLab/CRADLE/pull/78/files/c23ab4c9293eef8eb62c1aa1013162a071bbc9e9#diff-1fc2cff4e79bf185f9b94efbde101e4cd33c6b1d310d47459d05367f6fe0355cR126

Reply:
`selectReplicateIdx` could result in selecting different regions across samples, though?

Reply:
yes. I broke the old `selectIdx` function into three different functions -- `selectOverallIdx`, `selectReplicateIdx`, and `selectHighRCIdx` -- one function for each set of indexes that `selectIdx` returned. values calculated by `selectReplicateIdx` replace the `ctrlSpecificIdx` and `experiSpecificIdx` from the old function.

Reply:
oh I see. I thought the previous versions of CRADLE filtered the same regions across all replicates but it was not—there were `ctrlSpecificIdx` and `experiSpecificIdx`.

Reply:
Ah, I misspoke. the previous versions of CRADLE didn't return  a "selectOverallIdx" value from `selectIdx`, but it does calculate that value and then use it to create `ctrlSpecificIdx` and `experiSpecificIdx`.

I decoupled generating those "overall" indexes from creating the `selectReplicateIdx` value, but `selectReplicatedIdx` does use the "overall" index to create the per-replicat index in the same way that `selectIdx` used `idx` to generate the`ctrlSpecificIdx` and `experiSpecificIdx` indexes.